addons/font/bmfont.c

```diff
@@ -0,0 +1,498 @@
+   range->count++;
+   reallocate(range);
+   range->characters[range->count - 1] = data->c;
+}
+
+static void combine_ranges(BMFONT_DATA *data,
+      BMFONT_RANGE *range, BMFONT_RANGE *range2) {
+   (void)data;
+   range->count += range2->count;
+   reallocate(range);
+   memmove(range->characters + range->count - range2->count,
+      range2->characters,
+      range2->count * sizeof *range->characters);
+   range->next = range2->next;
+   al_free(range2->characters);
+   al_free(range2);
+}
+
+static void insert_new_range(BMFONT_DATA *data, BMFONT_RANGE *prev,
+      int codepoint) {
+   BMFONT_RANGE *range = al_calloc(1, sizeof *range);
+   range->first = codepoint;
+   range->count = 1;
+   reallocate(range);
+   range->characters[0] = data->c;
+   if (prev) {
+      range->next = prev->next;
+      prev->next = range;
+   }
+   else {
+      range->next = data->range_first;
+      data->range_first = range;
+   }
+}
+
+/* Adding a new codepoint creates either a new range consisting of that
+ * single codepoint, appends the codepoint to the end of a range,
+ * prepends it to the beginning of a range, or merges two ranges.
+ */
+static void add_codepoint(BMFONT_DATA *data, int codepoint) {
+   BMFONT_RANGE *prev = NULL;
+   BMFONT_RANGE *range = data->range_first;
+   while (range) {
+      if (codepoint == range->first - 1) {
+         prepend_char(data, range);
+         return;
+      }
+      if (codepoint < range->first) {
+         insert_new_range(data, prev, codepoint);
+         return;
+      }
+      if (codepoint == range->first + range->count) {
+         append_char(data, range);
+         BMFONT_RANGE *range2 = range->next;
+         if (range2 != NULL && codepoint == range2->first - 1) {
+            combine_ranges(data, range, range2);
+         }
+         return;
+      }
+      prev = range;
+      range = range->next;
+   }
+   insert_new_range(data, prev, codepoint);
+}
+
+static BMFONT_CHAR *find_codepoint(BMFONT_DATA *data, int codepoint) {
+   BMFONT_RANGE *range = data->range_first;
+   while (range) {
+      if (codepoint >= range->first &&
+            codepoint < range->first + range->count) {
+         return range->characters[codepoint - range->first];
+      }
+      range = range->next;
+   }
+   return NULL;
+}
+
+static void add_page(BMFONT_DATA *data, char const *filename) {
+   data->pages_count++;
+   data->pages = al_realloc(data->pages, data->pages_count *
+      sizeof *data->pages);
+   al_set_path_filename(data->path, filename);
+   ALLEGRO_BITMAP *page = al_load_bitmap_flags(
+      al_path_cstr(data->path, '/'), data->flags);
+   data->pages[data->pages_count - 1] = page;
+}
+
+static bool tag_is(BMFONT_DATA *data, char const *str) {
+   return strcmp(al_cstr(data->tag), str) == 0;
+}
+
+static bool attribute_is(BMFONT_DATA *data, char const *str) {
+   return strcmp(al_cstr(data->attribute), str) == 0;
+}
+
+static int get_int(char const *value) {
+   return strtol(value, NULL, 0);
```

Comment:
Why not force base 10?

Reply:
I always use hexadecimal numbers myself - in this case it doesn't really matter I guess.